### PR TITLE
CP-48385: Enhancements for xapi-guard cache

### PR DESCRIPTION
**Pull request by psafont**
This is a misc bag of enhancements:
- Increase capacity of the queue in the queue and starting state
- Protect better against invalid filesystem contents of the cache
- Avoid spamming logs
- Simplify in-cache file management

Best reviewed commit by commit

### CODE REVIEW

File: ocaml/xapi-guard/lib/disk_cache.ml [hunk 1]
```diff
@@ -585,19 +567,29 @@ end = struct
...
+    let* () = Lwt_list.iter_p (commit __FUNCTION__) actions in
+    let* () = delete_empty_dirs ~delete_root:false root in
+    Lwt.return (List.filter_map Fun.id pending)
 end
 
 let setup typ read write =
-  let* () = Setup.retime_cache_contents typ in
-  let queue, push = Lwt_bounded_stream.create 2 in
+  let* pending = Setup.retime_cache_contents typ in
+  let capacity = 4096 in
```
Comment (lindig) [inline ocaml/xapi-guard/lib/disk_cache.ml:580]: This looks like a rather large value for a list length.

File: ocaml/xapi-guard/lib/disk_cache.ml [hunk 2]
```diff
@@ -585,19 +567,29 @@ end = struct
...
+  let queue, push = Lwt_bounded_stream.create capacity in
   let lock = Lwt_mutex.create () in
-  let q = {queue; push; lock; state= Disengaged} in
+  let state =
+    if pending = [] then
+      Direct
+    else if List.length pending < capacity then
```
Comment (lindig) [inline ocaml/xapi-guard/lib/disk_cache.ml:586]: Here we are using non-tail recursive functions on a potentially large list. I believe 4096 is not causing a stack overflow but I do think this large enough to wonder if this is the right data structure.

File: ocaml/xapi-guard/lib/disk_cache.ml [hunk 3]
```diff
@@ -382,43 +371,38 @@ module Watcher : sig
...
-      List.filter_map (function Latest f -> Some f | _ -> None) keep
-    in
     Lwt.return latest
 
   let retry_push push (uuid, timestamp, key) contents =
     let __FUN = __FUNCTION__ in
     let push' () = push (uuid, timestamp, key) contents in
-    let rec retry k =
+    let counter = Mtime_clock.counter () in
+    let rec retry is_first_try =
```
Comment (lindig) [inline ocaml/xapi-guard/lib/disk_cache.ml:384]: I would suggest to make the argument a label.

**Review by lindig — APPROVED** [review]

File: ocaml/xapi-guard/lib/disk_cache.ml (hunk 2 above)
Comment (psafont) [reply ocaml/xapi-guard/lib/disk_cache.ml:586]: I'll try using sequences here, it's a bit involved since the lwt libraries like to use lists

Comment (psafont) [reply ocaml/xapi-guard/lib/disk_cache.ml:586]: I've changed this to an tail-recursive iter

File: ocaml/xapi-guard/lib/disk_cache.ml (hunk 1 above)
Comment (psafont) [reply ocaml/xapi-guard/lib/disk_cache.ml:580]: I've reduced this to 512, it should give less problems with the stack, while being big enough to handle most cases (512 VMs in a host with TPM that have written while restarting)

File: ocaml/xapi-guard/lib/types.ml [hunk 4]
```diff
@@ -28,13 +28,13 @@ module Tpm = struct
...
-        Perm
+        Ok Perm
     | 1 ->
-        Save
+        Ok Save
     | 2 ->
-        Volatile
+        Ok Volatile
     | s ->
-        Fmt.invalid_arg "Unknown TPM state key: %i" s
```
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/xapi-guard/lib/types.ml:37]: invalid_arg does not just log, it raises an exception, interrupting the fetching of all keys in the cache. Using the result allows to continue the collection while logging the errors.

File: ocaml/xapi-guard/lib/disk_cache.ml [hunk 5]
```diff
@@ -62,11 +62,7 @@ let unlink_safe file =
 
 type valid_file = t * string
 
-type file =
-  | Latest of valid_file
-  | Outdated of valid_file
-  | Temporary of string
-  | Invalid of string
+type file = Latest of valid_file | Outdated of valid_file | Invalid of string
```
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/xapi-guard/lib/disk_cache.ml:65]: temp files are files that have been written to the filessystem, but not atomically moved to the final location.

File: ocaml/xapi-guard/lib/disk_cache.ml (hunk 2 above)
Comment (robhoes) [reply ocaml/xapi-guard/lib/disk_cache.ml:586]: Did you? I see a map below.

**Review by robhoes — APPROVED** [review]
I think you still need to replace a `List.map` by a `List.iter`, but otherwise this looks good.

**Pull request merged**